src/mgr/MgrClient.cc

```diff
@@ -199,45 +199,45 @@ void MgrClient::send_report()
   pcc->with_counters([this, report](
         const PerfCountersCollection::CounterMap &by_path)
   {
-    bool const declared_all = (session->declared.size() == by_path.size());
-
-    if (!declared_all) {
-      for (const auto &i : by_path) {
-        auto path = i.first;
-        auto data = *(i.second);
-        
-        if (session->declared.count(path) == 0) {
-          PerfCounterType type;
-          type.path = path;
-          if (data.description) {
-            type.description = data.description;
-          }
-          if (data.nick) {
-            type.nick = data.nick;
-          }
-          type.type = data.type;
-          report->declare_types.push_back(std::move(type));
-          session->declared.insert(path);
-        }
+    ENCODE_START(1, 1, report->packed);
+    for (auto p = session->declared.begin(); p != session->declared.end(); ) {
+      if (by_path.count(*p) == 0) {
+	report->undeclare_types.push_back(*p);
+	ldout(cct,20) << __func__ << " undeclare " << *p << dendl;
+	p = session->declared.erase(p);
+      } else {
+	++p;
       }
     }
+    for (const auto &i : by_path) {
+      auto& path = i.first;
+      auto& data = *(i.second);
+
+      if (session->declared.count(path) == 0) {
+	ldout(cct,20) << __func__ << " declare " << path << dendl;
+	PerfCounterType type;
+	type.path = path;
+	if (data.description) {
+	  type.description = data.description;
+	}
+	if (data.nick) {
+	  type.nick = data.nick;
+	}
+	type.type = data.type;
+	report->declare_types.push_back(std::move(type));
```

Comment:
```
2017-03-22 07:46:59.606967 7f7833ac0700 -1 *** Caught signal (Aborted) **
 in thread 7f7833ac0700 thread_name:ms_dispatch

 ceph version 12.0.0-1716-g019d749 (019d7497e9577c4ffcdd8d4ece247dfa58c6fe81)
 1: (()+0x282872) [0x5615e6aa6872]
 2: (()+0x11390) [0x7f783ca86390]
 3: (gsignal()+0x38) [0x7f783ba17428]
 4: (abort()+0x16a) [0x7f783ba1902a]
 5: (__gnu_cxx::__verbose_terminate_handler()+0x16d) [0x7f783c35984d]
 6: (()+0x8d6b6) [0x7f783c3576b6]
 7: (()+0x8d701) [0x7f783c357701]
 8: (()+0x8d919) [0x7f783c357919]
 9: (()+0x1c5e45) [0x5615e69e9e45]
 10: (()+0x28cdbd) [0x5615e6ab0dbd]
 11: (DaemonPerfCounters::update(MMgrReport*)+0x711) [0x5615e6a1d701]
 12: (DaemonServer::handle_report(MMgrReport*)+0x3f4) [0x5615e6a22154]
 13: (DaemonServer::ms_dispatch(Message*)+0x60) [0x5615e6a26a20]
 14: (DispatchQueue::entry()+0xf4a) [0x5615e6d137ea]
 15: (DispatchQueue::DispatchThread::entry()+0xd) [0x5615e6bb1b7d]
 16: (()+0x76ba) [0x7f783ca7c6ba]
 17: (clone()+0x6d) [0x7f783bae882d]
 NOTE: a copy of the executable, or `objdump -rdS <executable>` is needed to interpret this.

--- begin dump of recent events ---
```
i have this in /a/kchai-2017-03-22_07:40:06-rados-wip-kefu-testing---basic-mira/932172/remote/mira069/log/ceph-mgr.x.log.gz with https://github.com/ceph/ceph/pull/14078